File: hush-providers/hush/providers/registry/auth_plugin.py
"""Auth resource plugin for ResourceHub.

Auto-registers Keycloak config classes and factory handlers with hush-core.
"""

from hush.core.registry import REGISTRY
from hush.providers.auth.config import KeycloakTokenConfig
from hush.providers.auth.factory import create_auth

_registered = False


def register():
    """Register Keycloak config class and factory handler."""
    global _registered
    if _registered:
        return

    REGISTRY.register(KeycloakTokenConfig, create_auth)
    _registered = True


def is_registered() -> bool:
    """Check if plugin has been registered."""
    return _registered


# Auto-register on import
register()

File: rust/operonx/src/providers/registry/auth_plugin.rs
//! Auth plugin registration.
//!
//! Mirrors Python [`operonx/providers/registry/auth_plugin.py`](../../../../../operonx/providers/registry/auth_plugin.py).

use std::sync::Arc;

use serde_json::Value;

use super::KeycloakResource;
use crate::core::exceptions::OperonError;
use crate::core::registry::{registry, ConfigDict, Factory, ResourceInstance};
use crate::providers::auth::{create_auth, KeycloakTokenConfig};

/// Register the keycloak category factory. Idempotent.
pub fn register() -> Result<(), OperonError> {
    if registry().get_factory("keycloak").is_some() {
        return Ok(());
    }
    let factory: Factory = Arc::new(|cfg: ConfigDict| {
        let typed: KeycloakTokenConfig = serde_json::from_value(Value::Object(cfg))?;
        let provider = create_auth(typed)?;
        Ok(Arc::new(KeycloakResource(provider)) as ResourceInstance)
    });
    registry().register("keycloak", factory, Some("KeycloakTokenConfig"))
}
